backend/cmd/headlamp.go

```diff
@@ -437,6 +657,131 @@ func StartHeadlampServer(config *HeadlampConfig) {
+	if err != nil {
+		return 0, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+
+	defer l.Close()
+
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+// nolint: funlen
+func (c *HeadlampConfig) startPortForward(p PortForwardPayload, token string) {
+	ports := []string{fmt.Sprintf(p.Port + ":" + p.TargetPort)}
+	ctxtProxy, ok := c.contextProxies[p.Cluster]
+
+	if !ok {
+		log.Fatalf("cluster %s not found", p.Cluster)
+		return
+	}
+
+	var caData []byte
+
+	var err error
+
+	if caData, err = ctxtProxy.context.cluster.getCAData(); err != nil {
+		log.Fatalf("error getting caData" + err.Error())
+		return
+	}
+
+	rConf := &rest.Config{
+		Host:        ctxtProxy.context.cluster.config.Server,
+		BearerToken: token,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData:   caData,
+			KeyFile:  ctxtProxy.context.authInfo.ClientKey,
+			CertFile: ctxtProxy.context.authInfo.ClientCertificate,
+			KeyData:  ctxtProxy.context.authInfo.ClientKeyData,
+			CertData: ctxtProxy.context.authInfo.ClientCertificateData,
+		},
+	}
+
+	fmt.Println("caData is", caData)
+	fmt.Println("\n\nkeyFile is", ctxtProxy.context.authInfo.ClientKey)
+	fmt.Println("\n\ncertFile is", ctxtProxy.context.authInfo.ClientCertificate)
+	roundTripper, upgrader, err := spdy.RoundTripperFor(rConf)
+	if err != nil {
+		log.Printf("Error: failed to create round tripper: %s", err)
+	}
+
+	requestURL := fmt.Sprintf("%s/api/v1/namespaces/%s/pods/%s/portforward", rConf.Host, p.Namespace, p.Pod)
+
+	reqURL, err := url.Parse(requestURL)
+	if err != nil {
+		log.Panicln("url parse err", requestURL, err)
+	}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, reqURL)
+	stopChan, readyChan := make(chan struct{}), make(chan struct{}, 1)
+	out, errOut := new(bytes.Buffer), new(bytes.Buffer)
+	forwarder, err := portforward.New(dialer, ports, stopChan, readyChan, out, errOut)
+	if err != nil {
+		log.Printf("Error: failed to create a portforward request: %s", err)
+	}
+
+	go func() {
+		if err = forwarder.ForwardPorts(); err != nil { // Locks until stopChan is closed.
+			log.Printf("Error: failed to forward ports: %s", err)
+		}
+	}()
+
+	for {
+		<-readyChan
+		break
+	}
+
+	if errOut.String() == "" {
+		p := PortForward{
+			ID:        p.ID,
+			closeChan: stopChan,
+			Pod:       p.Pod,
+			Cluster:   p.Cluster,
+			Namespace: p.Namespace,
+			Port:      p.Port,
+			Out:       out,
+			ErrOut:    errOut,
+		}
+		portforwardstore(p)
+	}
+
+	clientset, err := kubernetes.NewForConfig(rConf)
+	if err != nil {
+		log.Printf("kpoward: failed to create clientset: %s", err)
+	}
+	/* check every PodAvailabilityCheckTimer seconds if the pod for which we started a portforward is running
+	if not then we close the channel
+	*/
+	ticker := time.NewTicker(PodAvailabilityCheckTimer * time.Second)
+
+	go func() {
+		for range ticker.C {
+			pod, err := clientset.CoreV1().Pods(p.Namespace).Get(p.Pod, v1.GetOptions{})
+			if err != nil {
```

Comment:
This will stop the port forward even if Get pod fails due to network issues. Better error matching is required here.